### PR TITLE
Gloo backend support

### PR DESCRIPTION
**Pull request by Kacper-Pietkun**
- Added path for using all_to_all function with gloo backend (implemented with multiple isend calls)
- Added test for all_to_all function
- Updated examples to allow using gloo backend
- Updated docs with information about available backends

### CODE REVIEW

File: tests/test_comm.py
Comment (bgawrych) [inline tests/test_comm.py:1]: @pytest.mark.parametrize("backend", ["ccl", "gloo"])
We should parametrize all test with this one

File: docs/source/comm.rst [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 SAR's communication routines
 =============================
-SAR uses only two types of collective communication calls: ``all_to_all`` and ``all_reduce``. This choice was made to improve scalability by avoiding any point-to-point communication. Currently, the only backends in `torch.distributed <https://pytorch.org/docs/stable/distributed.html>`_ that support ``all_to_all`` are ``nccl``, ``ccl``, or ``mpi``.  Nvidia's ``nccl`` is already included in the PyTorch distribution and it is the natural choice when training on GPUs.
+SAR uses only two types of collective communication calls: ``all_to_all`` and ``all_reduce``. This choice was made to improve scalability by avoiding any point-to-point communication. SAR supports four backends, which are ``ccl``, ``nccl``, ``mpi`` and ``gloo``. Nvidia's ``nccl`` is already included in the PyTorch distribution and it is the natural choice when training on GPUs.
```
Comment (bgawrych) [inline docs/source/comm.rst:7]: Single sentence that 'gloo' doesn't support all_to_all and that a handcrafted implementation is used, which may not be as optimal as in other backends would be welcomed.

File: tests/test_comm.py [hunk 2]
```diff
@@ -77,3 +77,32 @@ def gather_grads(mp_dict, rank, world_size, tmp_dir):
...
+            sar.comm.all_to_all(recv_tensors_list, send_tensors_list)
+            mp_dict[f"result_{rank}"] = recv_tensors_list
+        except Exception as e: 
+            mp_dict["traceback"] = str(traceback.format_exc())
+            mp_dict["exception"] = e
+    
+    mp_dict = run_workers(all_to_all, world_size, backend=backend)
+    for rank in range(world_size):
+        for tensor in mp_dict[f"result_{rank}"]:
+            assert torch.all(torch.eq(tensor, torch.tensor([rank]*world_size)))
```
Comment (bgawrych) [inline tests/test_comm.py:108]:
```suggestion
            assert torch.all(torch.eq(tensor, torch.tensor([rank] * world_size)))
```

File: tests/test_comm.py
Comment (Kacper-Pietkun) [reply tests/test_comm.py:1]: Added `backend` parameter for all tests in test_comm.py (tests in other folders can be modified after refactoring them to use functions from base_utils.py)

**Comment by hesham-mostafa** [comment]
Thanks for adding gloo support. One thing that would have a large impact on performance is to modify the 'exchange_single_tensor' function in comm.py to use dist.isend and dist.recv if backend()=='gloo'. Currently 'exchange_single_tensor' uses all_to_all which is inefficient because we only want to exchange tensors between 2 machines. I had no other choice because 'ccl' didn't support dist.send and dist.recv.

**Comment by Kacper-Pietkun** [comment]
> Thanks for adding gloo support. One thing that would have a large impact on performance is to modify the 'exchange_single_tensor' function in comm.py to use dist.isend and dist.recv if backend()=='gloo'. Currently 'exchange_single_tensor' uses all_to_all which is inefficient because we only want to exchange tensors between 2 machines. I had no other choice because 'ccl' didn't support dist.send and dist.recv.

I have added a path in `exchange_single_tensor` function to use isend/recv when using gloo backend + tests. If you don't have any more comments, it is ready to merge

**Pull request merged**